### DISCUSSION TITLE
Android activities and GetBaseActivityEvents

### DISCUSSION THREAD

**Discussion by queequac** [unoplatform/uno#7992] [Q&A] [answered]
First of all, working with custom Android activities is not really documented. I am not sure if I have really a clue how this should be used within a Uno app.
Anyway, it says I could access to the main activity events by adding the following line, ideally in the App constructor:

` ApplicationViewHelper.GetBaseActivityEvents().Create += OnCreateEvent;
`

## On which page?
https://platform.uno/docs/articles/android-activities.html

## What's wrong?
But adding this line to the constructor and registering OnCreateEvent is enough so the app won't start anymore, throwing 
`System.Reflection.TargetInvocationException: 'Exception has been thrown by the target of an invocation.' instead.
`

## Any feedback?
Some guidance on Android activities would be nice, and maybe a hint how/where to register properly for the above event.

**Comment by jeromelaban** [comment, marked as answer]
Could you give a bit more context about what you're trying to do? Also, can you provide the inner exceptions for the `TargetInvocationException` ?

**Reply by queequac** [reply]
Sorry, did not really debug into it yesterday.
`ApplicationViewHelper.GetBaseActivityEvents()` returns null, so I cannot subscribe for OnCreateEvent of course. Mhm.

Aside from this, I have a service that returns an Intent with some Activity. (Usually the app would then open the UI of that activity.)
I am absolute lost how to do this scenario with Xamarin+Uno.
Looking through the documentation I found only the page referenced above talking about Activities.

In an ideal world my App would pass the current activity as an argument in the LaunchActivatedEventArgs, so I could simply look it up and navigate to the corresponding page. ;)

**Reply by jeromelaban** [reply]
ApplicationViewHelper should not be throwing indeed, it could be that the location in the documentation is incorrect. We'll look it up.

As for deep linking or protocol activation, maybe [this](https://platform.uno/docs/articles/features/protocol-activation.html?q=protocol#android) can help, though this is for Uri-based activation.

If you have a java based sample that explains what you're trying to do, we can try to find out a solution.

**Reply by queequac** [reply]
I am realizing an Authenticator, where I implement `AbstractAccountAuthenticator`.
When for example an account is going to be added, the authenticator's `AddAccount` method is called by the system. That's the place where usually an intent is created that points to an activity that provides the corresponding login/authentication user interface.

```
public override Bundle AddAccount(AccountAuthenticatorResponse response, string accountType, string authTokenType, string[] requiredFeatures, Bundle options)
{
     Intent intent = new Intent(_context, typeof(LoginActivity));
     intent.PutExtra(AccountManager.KeyAccountAuthenticatorResponse, response);
     Bundle bundle = new Bundle();
     bundle.PutParcelable(AccountManager.KeyIntent, intent);
     return bundle;
}
```

Right now I have no clue how to deal with activities other than the main activity when using Uno.
And when making the intent point to MainActivity instead (actually abusing it as the single entry point for everything), I cannot really detect that it is not a regular app start but some specific activity called with additional extras (see PutExtra and PutParcelable above).
Not even sure if this would still work when using URI activation instead or if the system demands using the AccountManager data to be used. Furthermore I have to return the activity's result via specific extras being put again to the intent that was used by the system when triggering the login activity.

Also aside from Accounts I'd be nosy how to implement activites for various scenarios in general, since they are often used for interactive workflows across app boundaries and not only for deep-linking to some different start page.

**Reply by queequac** [reply]
Seems I can solve my problem by implementing all activities as `ApplicationActivity` (only MainActivity sets `MainLauncher = true`), because then it always initializes the `App` with the UI and calls `OnLaunched` on the very first activity, no matter through which activity the app got actually started. In addition I use Xamarin.Essentials to access the `CurrentActivity` and to monitor activity events in general. Depending on the activity state and the actual activity type I simply navigate to the correct UIs.

But is it safe to use multiple activities derived from `ApplicationActivity`? I am not sure if this is intended use or if that one was designed for being the one and only main application entry.
Even though it looks promising, there might be side effects...?

**Reply by davidjohnoliver** [reply]
I suspect the answer is 'it depends'. It's probably ok and reasonable for your Android package to define different activities that each inherit from `ApplicationActivity` that have different system entry points.

However you may run into unexpected behaviour if you're somehow spawning multiple activities inheriting from `ApplicationActivity` within the same running process. To answer your question directly, `ApplicationActivity` definitely expects to be a singleton:

https://github.com/unoplatform/uno/blob/0c2e67f6e27ff872da7ba750ee94dbf7d45978d8/src/Uno.UI/UI/Xaml/ApplicationActivity.Android.cs#L28-L31

So don't be surprised if you encounter some strange issues in going this route.

**To the broader question of handling activities in general:** 

For activities that don't need to show XAML UI, you can safely inherit from the [`BaseActivity` class](https://github.com/unoplatform/uno/blob/master/src/Uno.UI/BaseActivity.Android.cs). If you just want to launch an intent, the [`DelegateActivity` class](https://github.com/unoplatform/uno/blob/master/src/Uno.UI/DelegateActivity.Android.cs) allows you to easily do so and await the results as a Task.

Launching multiple XAML-based activities within the same application instance, is not currently something Uno readily supports or is designed for; it's expected that cross-platform navigation idioms (like Frame navigation or Uno.Navigation) will be used instead.

**Reply by queequac** [reply]
Thanks @davidjohnoliver for the insights.
Actually I need multiple UI based activities in my case. But at the moment it seems to work even when re-entering a running process, so I will keep an eye on unexpected bevavior. 🙃
